### ISSUE TITLE
Check for available disk space before DB migration/re-encryption

### ISSUE THREAD

**Issue by IvanBelyakoff** [status-im/status-desktop#11501]
Starting 0.13 for the first time caused "re-encryption" to happen, which failed at the end due to insufficient disk space.

### Description

We should pre-calculate the available disk space to be sufficient before starting DB migration/re-encryption and not proceed until enough space.
![Status_DiskFull](https://github.com/status-im/status-desktop/assets/15627093/b7dee49b-f7e8-431e-9810-e2fdee4a9ee0)

**Comment by alexjba** [comment]
We should also check for available disk space on password change. It's the same process.

**Comment by anastasiyaig** [comment]
@jrainville it it something for your team?

**Comment by alexjba** [comment]
IMO we could improve the error message here to inform the user of how much disk space the app needs to complete the re-encryption.

We currently have the following  re-encryption flows that could fail:
1. Login - when we migrate the password from upper-case to lower-case
2. Login - when we migrate the sqlciper to V4
3. Password change

**Comment by jrainville** [comment]
> @jrainville it it something for your team?

Could be us or wallet. Or Alex since he has experience with DB re-encryption

**Comment by igor-sirotin** [comment]
cc @jrainville @alexjba 

I made some tests. Found an old `shared-account`, which was v3-encrypted.

# TLDR; Conclusions

1. The database was never corrupted.
Even after 2 failing re-encryptions I was able to free space and successfully re-encrypt the database.
3. The error message is distinctive enough.
Yes, it's a bit developer-vibe, since it's coming from go, but it contains enough details for the user.
4. For me it took not too much time to get the error (~30 seconds).

I think it's not worth now to fix this issue. 
1. There's no simple precise way to calculate required disk space.
It should be enough to `2*(<file.db>.size()`, but that's just an estimation. 
2. We could simply check if the disk has "roughly enough" space, but that doesn't block the user from occupying the disk space while the re-encryption is in progress.
3. We can't acquire the disk space before running the re-encryption. This is done by sqlite.
5. Would be nice the make the error message better. E.g. show how much disk space is needed. 
But that's not super-important now and also requires design. And also check clause (1).

The worst case that's possible here is that we successfully re-encrypt the main DB, but fail to re-encrypt the wallet DB. Then  the DBs will end up in conflict state. But:
1. There might be a different error, not just the disk full.
2. `Disk full` error is unlikely in this case, because main DB is usually larger than the wallet DB.

# Testing details

Input:
- Database size: ~5.1Gb. 
- During the test I never restarted the app.
- Last version of desktop the account dir was launched with: 54709cb1c (v3-encrypted)
- Current desktop version: cd3a1180b00dcc3ab248f02398061d8ca36cbd0b

Some commands I used for testing on MacOS:
```shell
# Create an file with 8 millions of blocks with 1k bytes block size, i.e. ~8Gb.
# There's a simpler command `mkfile`, but it doesn't occupy any actual space on disk.
> dd if=/dev/zero of=temp_1m bs=1k count=8m seek=0 

# Check free space on disk
> df -H

# Check file space occupied on disk
du -sh <filename>
```

## Attempt 1 - 76Mb free space (extremely low)

```shell
> df -H
Filesystem        Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk3s1s1    995G     15G     76M   100%    387k  739k   34%   /
```

1. Enter the password
2. Get _almost immediate_ error `NO SPACE LEFT ON DEVICE`.

```nim
One or more databases not created
ERR 2024-04-04 13:04:36.435+01:00 error:                                     topics="accounts-service" tid=1526368 file=service.nim:656 procName=verifyDatabasePassword errDesription="One or more databases not created"
DBG 2024-04-04 13:04:36.437+01:00 Account logged in                          topics="accounts-service" tid=1526368 file=service.nim:676
failed to open db failed to set `journal_mode` pragma: disk I/O error: no space left on device
ERROR[04-04|13:04:36.868|github.com/status-im/status-go/mobile/status.go:236] failed to update node config fleet       key-uid=0x6c1dbb3634e25a69cea733f9c9051a9aeda6019a432b1031cd326a7b9fed2a4e error="Failed to migrate db file: Failed to migrate v3 db to v4: failed to set `journal_mode` pragma: disk I/O error: no space left on device"
ERR 2024-04-04 13:04:36.869+01:00 node.login                                 topics="signals-manager" tid=1526368 file=signals_manager.nim:54 error="Failed to migrate db file: Failed to migrate v3 db to v4: failed to set `journal_mode` pragma: disk I/O error: no space left on device"
ERR 2024-04-04 13:04:36.869+01:00 login error                                topics="app-controller" tid=1526368 file=module.nim:458 methodName=onNodeLogin errDesription="Failed to migrate db file: Failed to migrate v3 db to v4: failed to set `journal_mode` pragma: disk I/O error: no space left on device"
```

<img width="1411" alt="Screenshot 2024-04-04 at 13 04 45" src="https://github.com/status-im/status-desktop/assets/25482501/16d5f54e-0bcb-4aeb-886f-ea923f11017c">

## Attempt 2 - 2Gb free space (half of DB size)

```shell
> dd if=/dev/zero of=temp_1m bs=1k count=8m seek=0
8388608+0 records in
8388608+0 records out
8589934592 bytes transferred in 63.340697 secs (135614778 bytes/sec)

> df -H
Filesystem        Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk3s1s1    995G     15G   2081M    88%    387k   20M    2%   /

```

1. Input the password
2. After _30 seconds_ get error `DATABASE OR DISK FULL`
3. There's still ~2Gb free space after this operation.

```nim
DBG 2024-04-04 13:08:36.577+01:00 primary_action                             topics="app-controller" tid=1526368 file=module.nim:207 currFlow=AppLogin currState=LoginKeycardEnterPassword
One or more databases not created
ERR 2024-04-04 13:08:36.585+01:00 error:                                     topics="accounts-service" tid=1526368 file=service.nim:656 procName=verifyDatabasePassword errDesription="One or more databases not created"
DBG 2024-04-04 13:08:36.585+01:00 Account logged in                          topics="accounts-service" tid=1526368 file=service.nim:676
ERROR[04-04|13:09:07.985|github.com/status-im/status-go/mobile/status.go:236] failed to update node config fleet       key-uid=0x6c1dbb3634e25a69cea733f9c9051a9aeda6019a432b1031cd326a7b9fed2a4e error="Failed to migrate db file: Failed to migrate v3 db to v4: database or disk is full"
ERR 2024-04-04 13:09:07.997+01:00 node.login                                 topics="signals-manager" tid=1526368 file=signals_manager.nim:54 error="Failed to migrate db file: Failed to migrate v3 db to v4: database or disk is full"
ERR 2024-04-04 13:09:07.998+01:00 login error                                topics="app-controller" tid=1526368 file=module.nim:458 methodName=onNodeLogin errDesription="Failed to migrate db file: Failed to migrate v3 db to v4: database or disk is full"
```

<img width="1411" alt="Screenshot 2024-04-04 at 13 22 30" src="https://github.com/status-im/status-desktop/assets/25482501/64a717b9-093e-4c02-aaea-60b5bad250e2">

## Attempt 3 - 11Gb free space (enough for re-encryption)

Free up space.

```shell
> df -H
Filesystem        Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk3s1s1    995G     15G     11G    59%    387k  104M    0%   /
```

1. Input the password
2. After 52 seconds: successful login

**Comment by alexjba** [comment]
Nice investigation!

> I think it's not worth now to fix this issue.

I agree with you! It's ok since the error message is saying there's no disk space (If the user is not intimidated by the first part of the error). Maybe we can work in that direction and improve the message.

BTW, we have the same flows on password change. That might happen more often than the migration, but we should see a similar error there.

**Comment by jrainville** [comment]
I moved this to 2.30 and reduced the priority, since the problem is only visual and does not corrupt the DBs

**Comment by iurimatias** [comment]
this needs to be reviewed in the context of the upcoming revamped onboardingthis needs to be reviewed in the context of the upcoming revamped onboarding

**Comment by jrainville** [comment]
Not worth the effort. The message already tells the problem and it's an OS issue anyway
